engines/agi/loader_a2.cpp

```diff
@@ -0,0 +1,496 @@
+	int volumeCount = volumeMap.size() / diskCount;
+	int disksFound = 1;
+	for (uint i = 1; i < imageFiles.size() && disksFound < diskCount; i++) {
+		uint imageFileIndex = (diskOneIndex + i) % imageFiles.size();
+		Common::Path &imageFile = imageFiles[imageFileIndex];
+
+		Common::SeekableReadStream *stream = fileMap[imageFile].createReadStream();
+		if (stream == nullptr) {
+			warning("AgiLoader_A2: unable to open disk image: %s", imageFile.baseName().c_str());
+			continue;
+		}
+
+		// image file size must be 140k
+		int32 fileSize = stream->size();
+		if (fileSize != A2_DISK_SIZE) {
+			delete stream;
+			continue;
+		}
+
+		// check each disk
+		for (int d = 1; d < diskCount; d++) {
+			// has disk already been found?
+			if (!_imageFiles[d].empty()) {
+				continue;
+			}
+
+			bool match = false;
+			for (int v = 0; v < volumeCount; v++) {
+				uint32 offset = volumeMap[(v * diskCount) + d];
+				if (offset == _EMPTY) {
+					continue;
+				}
+
+				// test for expected resource header
+				stream->seek(offset);
+				uint16 magic = stream->readUint16BE();
+				byte volume = stream->readByte();
+				uint16 size = stream->readUint16LE();
+				if (magic == 0x1234 && volume == v && stream->pos() + size <= stream->size()) {
+					match = true;
+				} else {
+					match = false;
+					break;
+				}
+			}
+
+			if (match) {
+				_imageFiles[d] = imageFile.baseName();
+				disksFound++;
+				break;
+			}
+		}
+
+		delete stream;
+	}
+
+	// populate _volumes with the locations of the ones we will use.
+	// for each volume, select the one on the first available disk.
+	_volumes.resize(volumeCount);
+	for (uint32 i = 0; i < volumeMap.size(); i++) {
+		int volume = i / diskCount;
+		int disk = i % diskCount;
+		if (volumeMap[i] != _EMPTY) {
+			// use this disk's copy of the volume
+			_volumes[volume].disk = disk;
+			_volumes[volume].offset = volumeMap[i];
+
+			// skip to next volume
+			i = ((volume + 1) * diskCount) - 1;
+		}
+	}
+}
+
+// returns disk count on success, 0 on failure
+int AgiLoader_A2::readDiskOne(Common::SeekableReadStream &stream, Common::Array<uint32> &volumeMap) {
+	// INITDIR is located at track 1, sector 3, for games that have it.
+	int diskCount;
+	bool success = true;
+	if (_vm->getGameID() == GID_KQ2) {
+		// KQ2 doesn't have INITDIR. Use known locations.
+		diskCount = A2_KQ2_DISK_COUNT;
+		success &= readDir(stream, A2_KQ2_LOGDIR_POSITION,   _logDir);
+		success &= readDir(stream, A2_KQ2_PICDIR_POSITION,   _picDir);
+		success &= readDir(stream, A2_KQ2_VIEWDIR_POSITION,  _viewDir);
+		success &= readDir(stream, A2_KQ2_SOUNDDIR_POSITION, _soundDir);
+		success &= readDir(stream, A2_KQ2_OBJECTS_POSITION,  _objects);
+		success &= readDir(stream, A2_KQ2_WORDS_POSITION,    _words);
+		// KQ2 doesn't have a volume map, probably because all the
+		// volumes on the data disks start at the first sector.
+		// Create one with known values so that it can also be
+		// be used for disk detection.
```

Comment:
```suggestion
		// used for disk detection.
```

Reply:
Well lookie here, we've got ourselves a reader...